include/libtorrent/ip_notifier.hpp

```diff
@@ -0,0 +1,81 @@
+/*
+
+Copyright (c) 2016, Steven Siloti
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+* Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+* Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in
+the documentation and/or other materials provided with the distribution.
+* Neither the name of the author nor the names of its
+contributors may be used to endorse or promote products derived
+from this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+POSSIBILITY OF SUCH DAMAGE.
+
+*/
+
+#ifndef TORRENT_IP_NOTIFIER_HPP_INCLUDED
+#define TORRENT_IP_NOTIFIER_HPP_INCLUDED
+
+#include <functional>
+#include <array>
+
+#include "libtorrent/error_code.hpp"
+#include "libtorrent/address.hpp"
+#include "libtorrent/io_service.hpp"
+
+#if defined TORRENT_BUILD_SIMULATOR
+#elif TORRENT_USE_NETLINK
+#include "libtorrent/netlink.hpp"
+#elif defined TORRENT_WINDOWS
+#include <boost/asio/windows/object_handle.hpp>
+#endif
+
+namespace libtorrent
+{
+	struct ip_change_notifier : boost::noncopyable
+	{
+		// cb will be invoked from within a thread calling ios.run()
+		// when a change is detected in the system's IP addresses
+		ip_change_notifier(io_service& ios, std::function<void()> cb);
+		~ip_change_notifier();
+
+		// start listening for changes
+		void start();
```

Comment:
the main reason I mention this is because I think this is such a useful utility that it may make sense to create a separate git repository for it and use it as a submodule. And for it to be generic, I think it would need a slightly different interface.